langchain/chains/combine_documents/map_reduce.py

```diff
@@ -141,31 +141,62 @@ def combine_docs(
         Combine by mapping first chain over all documents, then reducing the results.
         This reducing can be done recursively if needed (if there are many documents).
         """
-        results = self.llm_chain.apply(
-            # FYI - this is parallelized and so it is fast.
-            [{self.document_variable_name: d.page_content, **kwargs} for d in docs],
-            callbacks=callbacks,
-        )
-        return self._process_results(
-            results, docs, token_max, callbacks=callbacks, **kwargs
-        )
+        if len(docs) == 1:
+            results = [
+                {**{self.document_variable_name: docs[0].page_content}, **kwargs}
+            ]
+
+        else:
+            results = self.llm_chain.apply(
+                # FYI - this is parallelized and so it is fast.
+                [{self.document_variable_name: d.page_content, **kwargs} for d in docs],
+                callbacks=callbacks,
+            )
+
+        try:
+            import nest_asyncio
+
+            nest_asyncio.apply()
+            return asyncio.run(
+                self._aprocess_results(results, docs, callbacks=callbacks, **kwargs)
+            )
+        except ImportError:
+            raise ImportError(
```

Comment:
this is really common operation - we cant have this depend on optional imports. can we make this change in a separate pr? isnt related to title of PR of "do not run map on a single document"

Reply:
Thanks for taking the time!

True that, makes sense. I am trying to avoid code duplication of sync/async methods like _(a)process_results() but I guess there is no way around having two versions of them (like so often in the codebase) and possibly refactor them a little to minimize duplication. NOT having an async `_aprocess_results()` imo is not acceptable: Calling `await chain.arun()` sort of promises that things are running concurrently which would not be the case with a sync only method.

Reply:
I provided a version without the nest_asyncio dependency.